Update buildsystem to support customapp bundling

jcfr opened the pull request.

jcfr committed:
COMP: Fix configuration against Slicer by updating default C++ version

This commit updates default to account for Slicer/Slicer@dedb6c04d (COMP: Require at least C++17)

jcfr committed:
COMP: Specify VTK component using updated convention

This commit fixes warnings like the following:

  CMake Deprecation Warning at /path/to/Slicer-Release/VTK-build/lib/cmake/vtk-9.1/vtk-config.cmake:75 (message):
    The new name for the 'vtkIOLegacy' component is 'IOLegacy'

jcfr committed:
COMP: Remove obsolete integration of VTK_USE_FILE

This commit fixes the following warning:

  CMake Deprecation Warning at /path/to/Slicer-Release/VTK-build/lib/cmake/vtk-9.1/vtk-use-file-deprecated.cmake:1 (message):
    The `VTK_USE_FILE` is no longer used starting with 8.90.
  Call Stack (most recent call first):
    UKFTractography.cmake:34 (include)
    CMakeLists.txt:97 (include)

jcfr committed:
STYLE: Strip trailing spaces

jcfr committed:
STYLE: Convert tabs to spaces

jcfr committed:
COMP: Fix configuration updating ITKSetStandardCompilerFlags

File ITKSetStandardCompilerFlags copied from ITK v5.1b01

This commit fixes the following error:

  CMake Error at /path/to/Slicer-build/ITK/CMake/ITK_CheckCCompilerFlag.cmake:29 (check_c_compiler_flag):
    Unknown CMake command "check_c_compiler_flag".

Adapted from:
* NIRALUser/ShapePopulationViewer@53edc736d (Fix configuration updating ITKSetStandardCompilerFlags)
* NIRALUser/ShapePopulationViewer@aecbfec22 (COMP: Fix inconsistent dll linkage error on windows)

jcfr committed:
COMP: Update CMake/PreventIn* modules

Update PreventInSourceBuilds module to reference Github instead of SVN
repository.

Adapted from:
* NIRALUser/ShapePopulationViewer@e7c97d75c (COMP: Fix Qt5 build and update SuperBuild layout using ExternalProjectDependency)

jcfr committed:
COMP: Remove copy of ExternalData CMake module

This commit removes obsolete copy of the module originally introduced
in 2013 through 6ac1ecfee (COMP: Switch to NamicExternalProjects SuperBuild)
and instead rely on the module provided by CMake.

jcfr committed:
COMP: Remove unused NAMICExternalProjectsExternalData CMake module

jcfr commented:
@yrathi Once this is integrated, we will rebase & consolidate change associated with #143 and #146

cc: @pieper @LucasGlandel

pieper commented:
Thank you @jcfr 🙏 

@yrathi it's hard to review these changes but I think we should accept to keep the code current or it will fall far behind.  Old code and binaries will still be available in the case that something unusual is introduced.

jcfr commented:
The white space changes can be excluded from the review appending `?w=1` to the URL.

For example, see https://github.com/pnlbwh/ukftractography/pull/149/files?w=1

Or by explicitly toggling the option:

![image](https://user-images.githubusercontent.com/219043/212384007-f2e43b5c-cd30-43b3-bf6a-f0ba5fff0eb4.png)

References:
* https://github.blog/changelog/2021-10-14-hiding-whitespace-is-now-remembered-for-each-pull-request/

tashrifbillah commented:
Please wait ... we are dealing with failed storage at our office now.

LucasGandel commented:
Tested on Windows with VS2019. Works like a charm after adding the following changes:
[UKF_diff.txt](https://github.com/pnlbwh/ukftractography/files/10446447/UKF_diff.txt)

cc: @jcfr

tashrifbillah commented:
What GCC did you use to test it on CentOS 7? Are you able to build it using the native 4.8.5?

jcfr committed:
COMP: Update Boost external project to support VS 2019

See https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering
and https://github.com/scikit-build/cmake-FindVcvars/blob/v1.6/FindVcvars.cmake#L124

Co-authored-by: Lucas Gandel <lucas.gandel@kitware.com>

jcfr committed:
STYLE: Update Boost external project describing Visual Studio version

References:
* https://www.boost.org/doc/libs/1_81_0/tools/build/doc/html/index.html#bbv2.reference.tools.compiler.msvc
* https://github.com/scikit-build/cmake-FindVcvars/blob/v1.6/FindVcvars.cmake#L124
* https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering

jcfr committed:
COMP: Update Boost external project adding support for VS 2022

References:
* https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering
* https://github.com/scikit-build/cmake-FindVcvars/blob/v1.6/FindVcvars.cmake#L124

jcfr committed:
STYLE: Set Eigen repository and version in corresponding external project

Streamline maintenance ensuring each project repository and version are
set in the corresponding external project file.

jcfr committed:
COMP: Remove unused and obsolete UKFTractography_USE_GIT_PROTOCOL option

jcfr committed:
COMP: Ensure GIT_EXECUTABLE is passed to inner build

jcfr committed:
COMP: Include "Common.cmake" in top-level CMakeLists.txt

jcfr committed:
COMP: Execute PreventIn(BuildInstalls|SourceBuilds) check as early as possible

jcfr committed:
COMP: Simplify OSX Deployment target setting

Remove outdated comments and logic for old MacOSX platforms that
are no longer supported.

Adapted from Slicer/Slicer@ac191b40f

jcfr committed:
COMP: Require at least macOS 10.13

jcfr commented:
Can we move forward with the integration ?

tashrifbillah commented:
https://github.com/pnlbwh/ukftractography/pull/149#issuecomment-1387180444

comment unaddressed.

jcfr commented:
> What GCC did you use to test it on CentOS 7?

Testing was done using `GCC 7`, see latest environment described at https://github.com/Slicer/SlicerBuildEnvironment#supported-build-environments (itself based of https://github.com/dockbuild/dockbuild#compiling-environments)


> Are you able to build it using the native 4.8.5?

Instead of using the outdated `4.8.5` compiler on Centos 7, we recommend upgrading your build environment leveraging the RedHat developer toolset.

Doing so allows to use modern compiler while being being compatible with older GlibC.

tashrifbillah commented:
> we recommend upgrading your build environment leveraging the RedHat developer toolset.

Hard to follow this recommendation in our workstation and cluster environment. They are all CentOS 7 and GCC 4.8.5 based. We shall discuss it internally.

tashrifbillah commented:
In any case, does your PR supersede #143?
@hjmjohnson

jcfr commented:
As described in https://github.com/pnlbwh/ukftractography/pull/149#issuecomment-1381748939

> @yrathi Once this is integrated, we will rebase & consolidate change associated with #143 and #146

jcfr commented:
> Did you add any new tests that would confirm that your builds did not break the underlying algebra/arithmetic?

Since these set of changes only focus on updating the build-system so that it is more consistent and maintainable, there were no tests added.

tashrifbillah commented:
@jhlegarreta, does your PR agree/conflict with or is superseded by this PR?

jhlegarreta commented:
> @jhlegarreta, does your PR agree/conflict with or is superseded by this PR?

This PR does not supersede my PR in any ways: they do different things. My PR offers a way to check that any given PR to the repository (and hopefully any commit) does not introduce configuration or compilation failure, or a regression. That is, if my PR gets merged, for any given PR, the code will be compiled, existing tests will be executed and the result compared to a baseline; if the CI comes green, then we know the changes do not introduce a configuration or compilation failure or regression.

This PR changes, updates or modernizes the build system. That is independent of having a CI. The CI would easily tell us if these changes are fine (at least for the tested configuration).

tashrifbillah commented:
Building in progress ...
cc @dheshanm

dheshanm commented:
Building the current codebase as-is with cmake-3.23.0-rc3-linux-x86_64 and gcc-9.3.0 results in the following cmake errors:

```
-- Setting C++ standard
-- Setting C++ standard - C++17
CMake Error at CMake/ExternalProjectDependency.cmake:75 (message):
  Failed to initialize variable SUPERBUILD_TOPLEVEL_PROJECT.  Variable
  CMAKE_PROJECT_NAME is not defined.
Call Stack (most recent call first):
  Common.cmake:2 (include)
  CMakeLists.txt:97 (include)


-- Configuring incomplete, errors occurred!
```

#### Steps to reproduce:

1. git clone git@github.com:jcfr/ukftractography.git -b update-buildsystem-to-support-customapp-bundling
2. cd ukftractography
3. mkdir build && cd build
4. cmake ../

I can further confirm that the same configuration, with the same steps builds master branch at [pnlbwh/ukftractography](https://github.com/pnlbwh/ukftractography) without issue.

dheshanm commented on UKFTractography.cmake:
```diff
@@ -20,18 +19,16 @@ include(${ITK_USE_FILE})
+  IOLegacy
+  IOXML
+  REQUIRED
+  )
```
Reverting these changes back to:

```
set(VTK_FOUND OFF)
find_package(VTK COMPONENTS
  vtkCommonSystem
  vtkCommonCore
  vtkCommonSystem
  vtkCommonMath
  vtkCommonMisc
  vtkCommonTransforms
  vtkIOLegacy
  vtkIOXML
  REQUIRED)
if(VTK_USE_FILE)
  include(${VTK_USE_FILE})
endif()
```

jcfr commented on UKFTractography.cmake:
```diff
@@ -20,18 +19,16 @@ include(${ITK_USE_FILE})
 #-----------------------------------------------------------------------------
 set(VTK_FOUND OFF)
 find_package(VTK COMPONENTS
```
```suggestion
find_package(VTK 9 COMPONENTS
```

This will make thing unambiguous

dheshanm commented:
I was able to get the build to succeed by making the following changes:

1. Move `include(${CMAKE_CURRENT_SOURCE_DIR}/Common.cmake)` after project definition on CMakeLists.txt
2. Reverting changes to VTK on UKFTractography.cmake
3. Using 'git' instead of ${GIT_EXECUTABLE} on ukf/CMakeLists.txt

### More Details
<details>
  <summary>1. Move `include(${CMAKE_CURRENT_SOURCE_DIR}/Common.cmake)` after project definition on CMakeLists.txt </summary>
  
https://github.com/pnlbwh/ukftractography/blob/32ec799c6b12b10a0bd92fdfb5b66e1e5e992b02/CMakeLists.txt#L97-L108

Move `include(${CMAKE_CURRENT_SOURCE_DIR}/Common.cmake)` after project definition.

```
#-----------------------------------------------------------------------------
# Superbuild script
#-----------------------------------------------------------------------------

if(${PRIMARY_PROJECT_NAME}_SUPERBUILD)
  project(SuperBuild_${PRIMARY_PROJECT_NAME} LANGUAGES C CXX)  # <- NOTE: Project name for pre-requisites is different form main project

 # ===== MOVED HERE BELOW =====
  include(${CMAKE_CURRENT_SOURCE_DIR}/Common.cmake) #<-- All feature options for top superbuild and inner product build
 # ===== MOVED HERE ABOVE =====

  include("${CMAKE_CURRENT_SOURCE_DIR}/SuperBuild.cmake")
  return()
else()
  # Building against Slicer
  project(${PRIMARY_PROJECT_NAME} LANGUAGES C CXX)             # <- NOTE: Here is the main project name setting
  include("${CMAKE_CURRENT_SOURCE_DIR}/${PRIMARY_PROJECT_NAME}.cmake")
  return()
endif()
message(FATAL_ERROR "You should never reach this point !")
```
</details>


<details>
  <summary>2. Reverting changes to VTK on UKFTractography.cmake </summary>

https://github.com/pnlbwh/ukftractography/blob/32ec799c6b12b10a0bd92fdfb5b66e1e5e992b02/UKFTractography.cmake#L20-L31
 
- Reverting changes to VTK back to 

```
  set(VTK_FOUND OFF)
  find_package(VTK COMPONENTS
    vtkCommonSystem
    vtkCommonCore
    vtkCommonSystem
    vtkCommonMath
    vtkCommonMisc
    vtkCommonTransforms
    vtkIOLegacy
    vtkIOXML
    REQUIRED)
  if(VTK_USE_FILE)
    include(${VTK_USE_FILE})
  endif()
```


</details>

<details>
  <summary>3. Using 'git' instead of ${GIT_EXECUTABLE} on ukf/CMakeLists.txt </summary>

https://github.com/pnlbwh/ukftractography/blob/32ec799c6b12b10a0bd92fdfb5b66e1e5e992b02/ukf/CMakeLists.txt#L14
 
Instead using `COMMAND git rev-parse HEAD  >> "${UKF_GITVER_TEMP}"`

</details>

### Further

Will try building with the proposed `find_package(VTK 9 COMPONENTS` again.

tashrifbillah commented:
> Move include(${CMAKE_CURRENT_SOURCE_DIR}/Common.cmake) after project definition on CMakeLists.txt

@jcfr Jean, I think this is another major problem. Feel free to propose a solution.

dheshanm commented:
Tried building with proposed `find_package(VTK 9 COMPONENTS` results in the following build error:

```
CMake Error at UKFTractography.cmake:21 (find_package):
  Could not find a configuration file for package "VTK" that is compatible
  with requested version "9".

  The following configuration files were considered but not accepted:

    /build/VTK-build/VTKConfig.cmake, version: 7.1.1

Call Stack (most recent call first):
  CMakeLists.txt:110 (include)
```

Note: This build used the modifications 1 and 3 as outlined from the above comment, and used the proposed `find_package(VTK 9 COMPONENTS`  for the 2nd issue instead of reverting.

tashrifbillah commented:
> That makes sense since I initially tried to avoid updating VTK external project .. which is still building VTK 7.

Were you able to successfully build this PR before proposing it? @jcfr

pieper commented:
I'd be fine with whatever builds best on modern systems.

@jcfr is this still useful for you?

jcfr commented:
The changes are expected to be general and still relevant, while I actively working on this I had started to further improve the PR to address both the "standalone" and "bundled" case but had to pause due to other constraints.

Waiting I have time to address those, I will submit smaller PR with stylistic changes and alike.

tashrifbillah commented:
> That makes sense since I initially tried to avoid updating VTK external project .. which is still building VTK 7.
> 
> I will follow in the next few days with an update. Thanks for the review

@jcfr , this PR does not build on RHEL 9 either. I get Dheshan's error: https://github.com/pnlbwh/ukftractography/pull/149#issuecomment-1682323036

He already proposed his solution here: https://github.com/pnlbwh/ukftractography/pull/149#issuecomment-1688737191
Can you follow up with a revision?

tashrifbillah commented:
Hi all, we have successfully built UKFTractography using all latest dependencies. Feel free to re-work this PR off the latest master if you need to.

PR #169

tashrifbillah commented:
Closed due to inactivity.

jcfr commented:
Thanks again for considering these changes in the first place. 

Commits were all carefully crafted with meaningful messages to streamline review.

Would rebasing those help ?

jcfr commented:
cc: @ungi
